### PR TITLE
jsvObjectSetChildAndUnLock string constants to Flash for ESP8266

### PR DESCRIPTION
**Pull request by wilberforce**
https://github.com/espruino/Espruino/issues/697#issuecomment-208298721

The  `do { ... } while (0)` trick, reduced the changes quite considerably.

This yields about 600 bytes or so of heap - which goes a long way to reducing memory issues when you try to use web sockets on this board.

### CODE REVIEW

File: src/jsutils.h [hunk 1]
```diff
@@ -59,12 +59,28 @@ char *flash_strncpy(char *dest, const char *source, size_t cap);
     and extract just the 8 bits we want */
 #define READ_FLASH_UINT8(ptr) ({ uint32_t __p = (uint32_t)(char*)(ptr); volatile uint32_t __d = *(uint32_t*)(__p & (uint32_t)~3); ((uint8_t*)&__d)[__p & 3]; })
 
+#define __concat(o,l) __to_rom##l
+#define jsvObjectSetChildAndUnLock(obj,name,child) jsvObjectSetChildAndUnLock_FLASH_CTR(obj,__COUNTER__,name,child)
```
Comment (tve) [inline src/jsutils.h]: I don't understand why this is needed. What is the path from the strings in flash to here? Why are the changes to the JswSymPtrs not all contained in the generated jswrapper.c file?

**Comment by wilberforce** [comment]
@tve 
> Is there a reason you created another section as opposed to reusing the same as the FLASH_STR macro?

I was trying to minimise the changes to the generator code and had a conflict when the in the linker when the const keyword was used.

Using this:
`#define IROM_CONST __attribute__((section(".irom.literal"))) __attribute__((aligned(4)))`

````
LD espruino_esp8266_partial.o
targets/esp8266/esp8266_board_utils.c:44:1: error: __ESPCONN_MEM causes a section type conflict with jswSymbols_process
 FLASH_STR(__ESPCONN_MEM, "ESPCONN_MEM");
 ^
gen/jswrapper.c:1087:24: note: 'jswSymbols_process' was declared here
 static const JswSymPtr jswSymbols_process[] IROM_CONST = {
                        ^
lto-wrapper: xtensa-lx106-elf-gcc returned 1 exit status
````

This can be avoided by dropping the const keyword here:

`static const JswSymPtr jswSymbols_E[] IROM_CONST = {`

`static JswSymPtr jswSymbols_E[] IROM_CONST = {`

However this would either need a macro in `scripts/build_jswrapper.py` and I could not work out how to make the generator conditional on the board, and I thought @gfwilliams would prefer not to have conditional code for one board.

Do you know another way of avoiding `rror: __ESPCONN_MEM causes a section type conflict with jswSymbols_process` without creating a new section?

File: src/jsutils.h (hunk 1 above)
Comment (wilberforce) [inline src/jsutils.h]: This is a different change - separate from JswSymPtrs stuff.

It moves all of the string literals used in `jsvObjectSetChildAndUnLock` to flash. It saves about 600 bytes of heap.

**Comment by tve** [comment]
You do need the two section as I've discovered myself. Sorry for the question.

**Comment by gfwilliams** [comment]
@tve @wilberforce  I had a build error on the ESP8266 branch with the `const` keyword last week I think - the issue wasn't that stuff had the `const` keyword, but that *something else was being put into read-only memory that **didn't** have the const keyword*.

If you find what variable it is and whack  `const` on it, the issue should go away...

**Comment by wilberforce** [comment]
@gfwilliams Help! - I'm struggling here.

I want to revert the changes made at 95f5d97 so I can pull @tve changes made back in the master.

At this stage I'm ready to dump everything in this pull request and re-create it.

Is there a way to unpull and revert my fork before these changes?

**Comment by gfwilliams** [comment]
Well, you can definitely do `git cherry-pick` on a new branch in order to pull in your changes that *weren't* the one you wanted - there'll be a bunch of info online.

But - how much does memory this really save now? Surely now @tve's changes are in, the need to grab a few extra bytes isn't as high - I'm not a huge fan of what you've done with `ota.c` - it just seems like it's really obfuscated what's there for the sake of 100 bytes or so.

Next PR will be to change GetChildAndUnLock, then jsvNewFromString, then strcpy, and slowly the whole of Espruino's source is getting these random `#define ESP8266` all over it.

**Pull request closed without merging**